analysis.py

```diff
@@ -36,7 +37,56 @@ def visualize_clusters(directory, bins=None, plot_kwargs={}, save_kwargs={'dpi':
     fig.savefig(os.path.join(directory, 'electron-energy-vs-ion_energy-difference.png'), **save_kwargs)
     plt.close(fig)
 
-def optimal_angular_distribution_hyperparameters(train_data, save_dir, val_data, labels, train_indices, val_indices, L_max, bin_range):
+def optimal_k_means_hyperparameters(phi, data, train_data, val_data, train_indices, val_indices, cluster_range, L_max, num_bins, save_dir='.'):
+    print('Range of clusters to consider: ', cluster_range)
+    
+    parameters = []
+    k_labels = []
+
+    for N in cluster_range:
+        labels, _ = clustering.k_means_clustering(phi, num_clusters=N)
+        k_labels.append(labels)
+        labels_train = labels[train_indices]
+        labels_val = labels[val_indices]
+        
+        ion_data = []
+        for i in range(N):
+            ion_data.append(np.vstack((train_data[labels_train==i, 0:3], train_data[labels_train==i, 3:6])))
+        val_ion_data = np.vstack((val_data[:,0:3], val_data[:,3:6]))
+        val_ion_labels = np.hstack((labels_val, labels_val)).reshape(-1)
+        
+        electron_data = []
+        for i in range(N):
+            electron_data.append(np.vstack((train_data[labels_train==i, 9:12], train_data[labels_train==i, 12:15])))
+        val_elec_data = np.vstack((val_data[:,9:12], val_data[:,12:15]))
+        val_elec_labels = np.hstack((labels_val, labels_val)).reshape(-1)
+
+        Bs = []
+        for i in range(N):
+            B_lms, lm_order = fitting.fit_Y_lms_binning_least_squares(
+                electron_data[i], L_max,
+                num_bins,
+                only_even_Ls=False
```

Comment:
Why are we hard coding this value? I thought it depends on whether we converted to molecular frame or not.

Reply:
Aww I had something typed up here but I guess I forgot to send. My initial thought was to use the fact that the odd terms should be zero as an indicator of quality. After thinking it over again I'm not sure that idea would work in an iterative process between optimization of clustering parameters and the loss function. This is probably something that should be toggle-able in the config yml file.